src/main/java/org/candlepin/resource/util/CloudAuthTokenUtil.java

```diff
@@ -0,0 +1,209 @@
+/*
+ * Copyright (c) 2009 - 2023 Red Hat, Inc.
+ *
+ * This software is licensed to you under the GNU General Public License,
+ * version 2 (GPLv2). There is NO WARRANTY for this software, express or
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package org.candlepin.resource.util;
+
+import org.candlepin.auth.CloudAuthTokenType;
+import org.candlepin.auth.Principal;
+import org.candlepin.config.ConfigProperties;
+import org.candlepin.config.Configuration;
+import org.candlepin.config.ConversionException;
+import org.candlepin.pki.CertificateReader;
+import org.candlepin.util.Util;
+
+import org.keycloak.common.util.KeyUtils;
+import org.keycloak.crypto.Algorithm;
+import org.keycloak.crypto.AsymmetricSignatureSignerContext;
+import org.keycloak.crypto.KeyType;
+import org.keycloak.crypto.KeyUse;
+import org.keycloak.crypto.KeyWrapper;
+import org.keycloak.jose.jws.JWSBuilder;
+import org.keycloak.representations.JsonWebToken;
+
+import java.security.PrivateKey;
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+import java.util.Objects;
+
+import javax.inject.Inject;
+
+
+
+/**
+ * A utility class for generating cloud authentication tokens
+ */
+public class CloudAuthTokenUtil {
```

Comment:
This class should probably be in the org.candlepin.auth package, and called `CloudAuthTokenGenerator`, since it's not a utility, but key to cloud authentication. I know it's called by a *Resource class, but the only reason for that being a Resource is technical limitations (that the instance metadata is too large to fit in HTTP headers, so we had to create a new API endpoint for it).

Reply:
Updated